gpflow/optimizers/scipy.py

```diff
@@ -121,12 +126,14 @@ def unpack_tensors(to_tensors: Iterator[tf.Tensor], from_vector: tf.Tensor) -> L
         return values
 
     @staticmethod
-    def assign_tensors(to_tensors: Iterator[tf.Variable], values: Iterator[tf.Tensor]):
+    def assign_tensors(to_tensors: Iterable[tf.Variable], values: Iterable[tf.Tensor]) -> None:
         for tensor, tensor_vector in zip(to_tensors, values):
             tensor.assign(tensor_vector)
 
 
-def _compute_loss_and_gradients(loss_cb: LossClosure, variables: Variables):
+def _compute_loss_and_gradients(
+    loss_cb: LossClosure, variables: Variables
+) -> Tuple[tf.Tensor, tf.Tensor]:
```

Comment:
the return type of [`tf.GradientTape.gradient`](https://www.tensorflow.org/api_docs/python/tf/GradientTape#gradient) is

> a list or nested structure of Tensors (or IndexedSlices, or None), one for each element in sources. Returned structure is the same as the structure of sources.

Since `variables` is an `Iterable[tf.Tensor]`, how do we know what the type of `grads` is? We know it's the same as `variables`, but do we know more? Would it not be

```python3
V = TypeVar("V", bound=Variables)

def _compute_loss_and_gradients(
    loss_cb: LossClosure, variables: Variables
) -> Tuple[tf.Tensor, V]:
	...
```
though that seems overkill here. Since it's only used with a `Tuple[tf.Tensor, ...]`, we could be more restrictive in the optimizer and use that instead of `Iterable[tf.Tensor]`. Then we wouldn't need the TypeVar and can just have
```python3
def _compute_loss_and_gradients(
    loss_cb: LossClosure, variables: Tuple[tf.Tensor, ...]
) -> Tuple[tf.Tensor, Tuple[tf.Tensor, ...]]:
```

Reply:
I'd prefer the TypeVar, I find your second suggestion hard to read, and when it gets to the point that it gets in the way of understanding what's going on for the human I'd rather remove the type annotation altogether...